src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -0,0 +1,303 @@
+        /// <param name="isRootObject">True if the <paramref name="objects"/> are root or top-level objects, i.e., they were passed directly by the client, otherwise false.</param>
+        /// <param name="objects">Objects of the top-level type we want to perform a deep insert on.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool isRootObject, params T[] objects)
+        {
+            foreach (object obj in objects)
+            {
+                EntityDescriptor topLevelDescriptor = this.RequestInfo.EntityTracker.GetEntityDescriptor(obj);
+
+                // If we do not find an EntityDescriptor in the entitytracker
+                // for any of the provided objects then we throw an exception. 
+                if (topLevelDescriptor == null)
+                {
+                    throw Error.InvalidOperation(Strings.Context_EntityNotContained);
+                }
+
+                // If it's a top-level object sent by the user.
+                if (isRootObject)
+                {
+                    // Validate that we can only have one top level entity in Deep Insert.
+                    if (bulkUpdateGraph.TopLevelDescriptors.Count > 0)
+                    {
+                        throw Error.InvalidOperation(Strings.Context_DeepInsertOneTopLevelEntity);
+                    }
+
+                    bulkUpdateGraph.AddTopLevelDescriptor(topLevelDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor is EntityDescriptor entityDescriptor)
+                    {
+                        if (entityDescriptor.ParentEntity != null &&
+                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) &&
+                            !bulkUpdateGraph.Contains(entityDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)
+                            {
+                                throw Error.InvalidOperation(Strings.Context_DeepInsertDeletedOrModified);
+                            }
+
+                            bulkUpdateGraph.AddRelatedDescriptor(topLevelDescriptor, entityDescriptor);
+                            this.BuildDescriptorGraph(descriptors, false, entityDescriptor.Entity);
+                        }
+                        else
+                        {
+                            // We check if the link representation of the descriptor exists in the entity tracker and it is unchanged. 
+                            LinkDescriptor linkDescriptor = this.RequestInfo.EntityTracker.Links.Where(
+                                a => a.Source == topLevelDescriptor.Entity &&

```

Comment:
I think this check is not necessary in a `DeepInsert`. So in a deep insert I believe users have to call either `AddLink` or `SetLink` in order to link a resource to an existing link. If they don't do that I don't think we need to worry ourselves with whatever they fetched using `$expand`

Reply:
Got it.
Fixed

Reply:
```csharp
var customer = new Customer
{
    Id = 999,
    Name = "Customer 999",
    Age = 50
};

var customer2 = context.Customers.Expand("Orders").FirstOrDefault();
var order1 = customer2.Orders.FirstOrDefault();

context.AttachTo("Customers", customer);
context.AddLink(customer, "Orders", order1);

context.DeepInsert<Customer>(customer);
```